std-bits/base/src/main/java/org/enso/base/FileLineReader.java

```diff
@@ -0,0 +1,367 @@
+package org.enso.base;
+
+import com.ibm.icu.text.Normalizer2;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.enso.base.arrays.LongArrayList;
+import org.enso.polyglot.common_utils.Core_Text_Utils;
+import org.graalvm.polyglot.Context;
+
+public class FileLineReader {
+  public static class ByteArrayOutputStreamWithContains extends ByteArrayOutputStream {
+    public ByteArrayOutputStreamWithContains(int size) {
+      super(size);
+    }
+
+    // ** Creates a preloaded stream from a byte array. */
+    public static ByteArrayOutputStreamWithContains fromByteArray(byte[] bytes) {
+      var stream = new ByteArrayOutputStreamWithContains(0);
+      stream.buf = bytes;
+      stream.count = bytes.length;
+      return stream;
+    }
+
+    public boolean contains(byte[] bytes) {
+      if (bytes.length > count) {
+        return false;
+      }
+      for (int i = 0; i < count - bytes.length; i++) {
+        boolean found = true;
+        for (int j = 0; j < bytes.length; j++) {
+          if (buf[i + j] != bytes[j]) {
+            found = false;
+            break;
+          }
+        }
+        if (found) {
+          return true;
+        }
+      }
+      return false;
+    }
```

Comment:
As discussed, we probably could try using the https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm for slightly more efficient loop here.

Reply:
Also, I think to satisfy the interruptability guarantees, we should add `context.safepoint()` in the inner loops everywhere where the loop is not bounded by a small constant.

As discussed in #7183 the safepoints have negligible impact on performance, but they make the code interruptible, so we should try including them in our library code everywhere where a process may take a longer amount of time and need to be cancellable.